packages/boot/src/booters/component-application.booter.ts

```diff
@@ -0,0 +1,66 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/boot
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  Application,
+  Binding,
+  BindingFilter,
+  Constructor,
+  CoreBindings,
+  createBindingFromClass,
+  inject,
+} from '@loopback/core';
+import {Bootable, Booter, booter} from '../types';
+
+/**
+ * Create a booter for an application that exposes a component to be mounted
+ * to the main application.
+ * @param componentApp An instance of the component application
+ * @param filter Binding filter to selected bindings to be added. Default to all
+ * bindings.
+ */
+export function createBooterForComponentApplication(
+  componentApp: Application & Bootable,
+  filter: BindingFilter = () => true,
+): Constructor<Booter> {
+  /**
+   * A booter to boot artifacts for the component application
+   */
+  @booter('componentApplications')
+  class ComponentApplicationBooter implements Booter {
+    constructor(
+      @inject(CoreBindings.APPLICATION_INSTANCE) private mainApp: Application,
+    ) {}
+
+    async load() {
+      await componentApp.boot();
+      /**
+       * Add bindings from the component application to the main application
+       */
+      const bindings = componentApp.find(binding => {
+        return filter(binding) && !this.mainApp.contains(binding.key);
```

Comment:
A question for `!this.mainApp.contains(binding.key)`:
what if the sub app wants to override the main app's bindings?

Reply:
There are certain bindings from the sub-app should not override the main app. We should carefully review such bindings - for example, `CoreBindings.*`, `RestBindings.*`. Maybe one way to solve it is to mark such bindings locked. If so, we can update this code to only skip locked bindings in the main app.

Reply:
> We should carefully review such bindings - for example, `CoreBindings.*`, `RestBindings.*`

True, agreed. I came up with this question due to the chat happened on public slack this morning, I think that community user tries to contribute custom artifacts(like models, in his case) from the component.

> Maybe one way to solve it is to mark such bindings locked. If so, we can update this code to only skip locked bindings in the main app.

I feel it's a case by case preference, maybe we can provide a default filter to skip those core bindings. And if user really wants to override, they provide custom filters :)

(don't have to address it in this PR)

Reply:
I improved the binding filtering as follows:

- Bindings that exist in the component application before `boot` are skipped.
- Locked bindings in the main application will not be overridden.
- Other bindings are selected by the binding filter. By default, all bindings are picked.